File: engine.py
"""
Motor de pontuação Conclave MR — funções puras (config + dados → totais e detalhes).
Manter em sincronia com web/engine.js
"""

from __future__ import annotations

import unicodedata
from collections import defaultdict
from typing import Any

MedalKey = str  # "ou" | "pt" | "br"


def truthy(v: Any) -> bool:
    if v is None or v is False:
        return False
    s = str(v).strip().lower()
    return s in ("x", "1", "sim", "s", "true", "verdadeiro", "ok", "✓")


def num(v: Any, default: float = 0.0) -> float:
    if v is None or v == "":
        return default
    try:
        return float(v)
    except (TypeError, ValueError):
        return default


def pontos_participacao(row: dict[str, Any], pesos: dict[str, float]) -> float:
    b = num(row.get("inscricao"), 0)
    if isinstance(row.get("inscricao"), bool):
        b = 1.0 if row.get("inscricao") else 0.0
    c = num(row.get("pontualidade"), 0)
    if isinstance(row.get("pontualidade"), bool):
        c = 1.0 if row.get("pontualidade") else 0.0
    d = int(num(row.get("mr_total"), 0))
    e = int(num(row.get("mr_camisa"), 0))
    f = int(num(row.get("mr_biblia"), 0))
    g = int(num(row.get("visitantes"), 0))
    anim = truthy(row.get("animacao"))

    if (b + d) <= 0:
        return 0.0

    total = b * pesos["inscricao"] + c * pesos["pontualidade"]
    if d > 0 and e == d:
        total += pesos["uniforme"]
    if d > 0 and f == d:
        total += pesos["biblia"]
    total += g * pesos["visitante"]
    if anim:
        total += pesos["animacao"]
    return total


def pontos_punicoes(row: dict[str, Any], pesos: dict[str, float]) -> float:
    p = 0.0
    if truthy(row.get("mau_comportamento")):
        p += pesos["mau_comportamento"]
    return p


def normalize_match_title(s: Any) -> str:
    """Título em minúsculas sem acentos, para reconhecer provas de desempate."""
    t = unicodedata.normalize("NFD", str(s or "").strip().lower())
    return "".join(c for c in t if unicodedata.category(c) != "Mn")


def tiebreak_prova_bucket(titulo: str) -> str | None:
    """
    Agrupa provas pelos critérios de desempate.
    «Organização» antes de «Bíblia» — ambas começam por «Conhecimentos gerais».
    """
    t = normalize_match_title(titulo)
    if "conhecimentos gerais" in t and "organiz" in t:
        return "cg_org"
    if "conhecimentos gerais" in t and "bibl" in t:
        return "cg_bib"
    if "debate" in t and "versicul" in t:
        return "debate"
    return None


def gincana_pts_on_prova(
    places: dict[str, Any] | None,
    medalhas: dict[str, float],
    igreja_id: str,
) -> float:
    if not isinstance(places, dict):
        return 0.0
    pts = 0.0
    for key in ("ou", "pt", "br"):
        entry = places.get(key) or {}
        if not isinstance(entry, dict):
            continue
        if entry.get("igrejaId") == igreja_id and key in medalhas:
            pts += float(medalhas[key])
    return pts


def build_tiebreak_by_igreja(evento: dict[str, Any], dados: dict[str, Any]) -> dict[str, dict[str, float]]:
    """Desempate: contagem ou/pt e pontos de gincana em provas-chave (por título)."""
    igrejas = evento.get("igrejas") or []
    medalhas = evento.get("medalhas") or {"ou": 0, "pt": 0, "br": 0}
    podium = dados.get("podium") or {}
    igreja_ids = {g["id"] for g in igrejas}
    medal_counts = contar_medalhas_por_igreja(podium, igreja_ids)
    out: dict[str, dict[str, float]] = {}
    for g in igrejas:
        gid = g["id"]
        m = medal_counts.get(gid) or {"ou": 0, "pt": 0, "br": 0}
        cg_bib = debate = cg_org = 0.0
        for p in evento.get("provas") or []:
            bucket = tiebreak_prova_bucket(str(p.get("titulo") or ""))
            if not bucket:
                continue
            pl = podium.get(p["id"])
            pts = gincana_pts_on_prova(pl, medalhas, gid)
            if bucket == "cg_bib":
                cg_bib += pts
            elif bucket == "debate":
                debate += pts
            elif bucket == "cg_org":
                cg_org += pts
        out[gid] = {
            "ou": float(m.get("ou") or 0),
            "pt": float(m.get("pt") or 0),
            "cgBiblia": cg_bib,
            "debate": debate,
            "cgOrg": cg_org,
        }
    return out


def gincana_por_igreja(
    podium: dict[str, dict[str, Any]],
    medalhas: dict[str, float],
    igreja_ids: set[str],
) -> dict[str, float]:
    """Soma pontos de gincana por igreja a partir do pódio (cada prova → ou/pt/br)."""
    tot: dict[str, float] = defaultdict(float)
    for _prova_id, places in podium.items():
        if not isinstance(places, dict):
            continue
        for key in ("ou", "pt", "br"):
            entry = places.get(key) or {}
            if not isinstance(entry, dict):
                continue
            iid = entry.get("igrejaId")
            if iid and iid in igreja_ids and key in medalhas:
                tot[str(iid)] += float(medalhas[key])
    return dict(tot)


def pontuacao_extra(row: dict[str, Any]) -> float:
    v = row.get("pontuacao_extra")
    if v is None:
        v = row.get("embaixadas")
    return num(v, 0.0)


def compute_totals(evento: dict[str, Any], dados: dict[str, Any]) -> dict[str, Any]:
    """
    evento: meta, pesos, medalhas, igrejas[], provas[]
    dados: participacao{ igrejaId: row }, podium{ provaId: { ou/pt/br: { igrejaId, competidor } } }
    """
    pesos = evento["pesos"]
    medalhas = evento["medalhas"]
    igrejas = evento["igrejas"]
    igreja_ids = {g["id"] for g in igrejas}
    participacao = dados.get("participacao") or {}
    podium = dados.get("podium") or {}

    ginc_totals = gincana_por_igreja(podium, medalhas, igreja_ids)

    detalhes = []
    for g in igrejas:
        gid = g["id"]
        row = participacao.get(gid) or {}
        p_part = pontos_participacao(row, pesos)
        p_puni = pontos_punicoes(row, pesos)
        p_ginc = float(ginc_totals.get(gid, 0.0))
        p_ext = pontuacao_extra(row)
        total = p_part + p_puni + p_ginc + p_ext
        detalhes.append(
            {
                "igrejaId": gid,
                "igreja": g["nome"],
                "participacao": p_part,
                "punicoes": p_puni,
                "gincana": p_ginc,
                "pontuacao_extra": p_ext,
                "total": total,
            }
        )

    totais = [d["total"] for d in detalhes]
    ranks = rank_list(totais)

    medalhas_por_igreja = contar_medalhas_por_igreja(podium, igreja_ids)
    tiebreak_by_igreja = build_tiebreak_by_igreja(evento, dados)

    return {
        "detalhes": detalhes,
        "totais": totais,
        "ranks": ranks,
        "medalhasPorIgreja": medalhas_por_igreja,
        "gincanaPorIgreja": ginc_totals,
        "tiebreakByIgreja": tiebreak_by_igreja,
    }


def rank_list(totais: list[float]) -> list[int]:
    n = len(totais)
    out = [0] * n
    order = sorted(range(n), key=lambda i: totais[i], reverse=True)
    for j, i in enumerate(order):
        if j == 0:
            out[i] = 1
        else:
            prev = order[j - 1]
            out[i] = out[prev] if totais[i] == totais[prev] else j + 1
    return out


def contar_medalhas_por_igreja(
    podium: dict[str, dict[str, Any]],
    igreja_ids: set[str],
) -> dict[str, dict[str, int]]:
    """Conta ou/pt/br conquistados por igreja (para relatório)."""
    out: dict[str, dict[str, int]] = {iid: {"ou": 0, "pt": 0, "br": 0} for iid in igreja_ids}
    for places in podium.values():
        if not isinstance(places, dict):
            continue
        for key in ("ou", "pt", "br"):
            entry = places.get(key) or {}
            if not isinstance(entry, dict):
                continue
            iid = entry.get("igrejaId")
            if iid and iid in out and key in out[iid]:
                out[iid][key] += 1
    return out


def classificacao_ordenada(
    detalhes: list[dict[str, Any]],
    ranks: list[int],
    tiebreak_by_igreja: dict[str, dict[str, Any]] | None = None,
) -> list[dict[str, Any]]:
    """
    Ordena por total desc.; empates: ouro → prata → pts gincana em Conhecimentos Gerais da Bíblia
    → Debate de Versículos → Conhecimentos Gerais da Organização → nome.
    """
    tb = tiebreak_by_igreja or {}

    def sort_key(i: int) -> tuple:
        d = detalhes[i]
        gid = d["igrejaId"]
        t = tb.get(gid) or {}
        ou = float(t.get("ou") or 0)
        pt = float(t.get("pt") or 0)
        cg_bib = float(t.get("cgBiblia") or 0)
        deb = float(t.get("debate") or 0)
        cg_org = float(t.get("cgOrg") or 0)
        nome = str(d.get("igreja") or "")
        return (-d["total"], -ou, -pt, -cg_bib, -deb, -cg_org, nome)

    indexed = sorted(range(len(detalhes)), key=sort_key)
    return [{**detalhes[i], "posicao": ranks[i]} for i in indexed]


def validate_evento_minimal(evento: dict[str, Any]) -> list[str]:
    """Retorna lista de erros (vazia = OK)."""
    errs: list[str] = []
    for k in ("meta", "pesos", "medalhas", "igrejas", "provas"):
        if k not in evento:
            errs.append(f"Falta chave: {k}")
    if errs:
        return errs
    ids = [g["id"] for g in evento["igrejas"]]
    if len(ids) != len(set(ids)):
        errs.append("IDs de igrejas duplicados")
    pids = [p["id"] for p in evento["provas"]]
    if len(pids) != len(set(pids)):
        errs.append("IDs de provas duplicados")
    return errs


def empty_dados_template(igreja_ids: list[str], prova_ids: list[str]) -> dict[str, Any]:
    part = {
        gid: {
            "inscricao": True,
            "pontualidade": True,
            "mr_total": 0,
            "mr_camisa": 0,
            "mr_biblia": 0,
            "visitantes": 0,
            "animacao": False,
            "mau_comportamento": False,
            "pontuacao_extra": 0,
        }
        for gid in igreja_ids
    }
    pod = {
        pid: {"ou": {"igrejaId": None, "competidor": ""}, "pt": {"igrejaId": None, "competidor": ""}, "br": {"igrejaId": None, "competidor": ""}}
        for pid in prova_ids
    }
    return {"participacao": part, "podium": pod}

File: web/engine.js
/**
 * Motor de pontuação Pontuação Conclave (uso web).
 */
(function (global) {
  function truthy(v) {
    if (v === null || v === false) return false;
    const s = String(v).trim().toLowerCase();
    return ["x", "1", "sim", "s", "true", "verdadeiro", "ok", "✓"].includes(s);
  }

  function num(v, def) {
    if (def === undefined) def = 0;
    if (v === null || v === undefined || v === "") return def;
    // Aceita vírgula decimal (entrada de planilha em pt-BR) sem quebrar JSON.
    if (typeof v === "string") {
      const s = v.trim();
      if (s === "") return def;
      const n = Number(s.indexOf(",") !== -1 ? s.replace(/\./g, "").replace(",", ".") : s);
      return Number.isFinite(n) ? n : def;
    }
    const n = Number(v);
    return Number.isFinite(n) ? n : def;
  }

  /** Lê peso garantindo número finito (default 0) — evita NaN silencioso quando
   *  o evento foi importado com `pesos` parciais. */
  function peso(pesos, key) {
    if (!pesos) return 0;
    const v = pesos[key];
    return typeof v === "number" && Number.isFinite(v) ? v : 0;
  }

  function pontosParticipacao(row, pesos) {
    let b = num(row.inscricao, 0);
    if (typeof row.inscricao === "boolean") b = row.inscricao ? 1 : 0;
    let c = num(row.pontualidade, 0);
    if (typeof row.pontualidade === "boolean") c = row.pontualidade ? 1 : 0;
    const d = Math.trunc(num(row.mr_total, 0));
    const e = Math.trunc(num(row.mr_camisa, 0));
    const f = Math.trunc(num(row.mr_biblia, 0));
    const g = Math.trunc(num(row.visitantes, 0));
    const anim = truthy(row.animacao);
    if (b + d <= 0) return 0;
    let total = b * peso(pesos, "inscricao") + c * peso(pesos, "pontualidade");
    if (d > 0 && e === d) total += peso(pesos, "uniforme");
    if (d > 0 && f === d) total += peso(pesos, "biblia");
    total += g * peso(pesos, "visitante");
    if (anim) total += peso(pesos, "animacao");
    return total;
  }

  function pontosPunicoes(row, pesos) {
    let p = 0;
    if (truthy(row.mau_comportamento)) p += peso(pesos, "mau_comportamento");
    return p;
  }

  function gincanaPorIgreja(podium, medalhas, igrejaIds) {
    const tot = {};
    igrejaIds.forEach(function (iid) {
      tot[iid] = 0;
    });
    Object.values(podium).forEach(function (places) {
      if (!places || typeof places !== "object") return;
      ["ou", "pt", "br"].forEach(function (key) {
        const entry = places[key];
        if (!entry || typeof entry !== "object") return;
        const iid = entry.igrejaId;
        if (iid && igrejaIds.has(iid) && medalhas[key] != null) {
          tot[iid] = (tot[iid] || 0) + Number(medalhas[key]);
        }
      });
    });
    return tot;
  }

  function pontuacaoExtra(row) {
    const v = row.pontuacao_extra != null ? row.pontuacao_extra : row.embaixadas;
    return num(v, 0);
  }

  /** Título normalizado para reconhecer provas de desempate (acentos ignorados). */
  function normalizeMatchTitle(s) {
    return String(s || "")
      .trim()
      .toLowerCase()
      .normalize("NFD")
      .replace(/[\u0300-\u036f]/g, "");
  }

  /**
   * Agrupa provas pelos critérios de desempate (título).
   * Ordem de teste: Organização antes de Bíblia (ambas «Conhecimentos gerais…»).
   */
  function tiebreakProvaBucket(titulo) {
    const t = normalizeMatchTitle(titulo);
    if (t.indexOf("conhecimentos gerais") !== -1 && t.indexOf("organiz") !== -1) return "cgOrg";
    if (t.indexOf("conhecimentos gerais") !== -1 && t.indexOf("bibl") !== -1) return "cgBib";
    if (t.indexOf("debate") !== -1 && t.indexOf("versicul") !== -1) return "debate";
    return null;
  }

  function gincanaPtsOnProva(places, medalhas, igrejaId) {
    if (!places || typeof places !== "object") return 0;
    let pts = 0;
    ["ou", "pt", "br"].forEach(function (key) {
      const entry = places[key];
      if (!entry || typeof entry !== "object") return;
      if (entry.igrejaId === igrejaId && medalhas[key] != null) {
        pts += Number(medalhas[key]);
      }
    });
    return pts;
  }

  function buildTiebreakByIgreja(evento, dados) {
    const igrejas = Array.isArray(evento && evento.igrejas) ? evento.igrejas : [];
    const medalhas = (evento && evento.medalhas) || { ou: 0, pt: 0, br: 0 };
    const podium = (dados && dados.podium) || {};
    const igrejaIds = new Set(
      igrejas.map(function (g) {
        return g && g.id;
      })
    );
    const medalCounts = contarMedalhasPorIgreja(podium, igrejaIds);
    const out = {};
    igrejas.forEach(function (g) {
      const gid = g.id;
      const m = medalCounts[gid] || { ou: 0, pt: 0, br: 0 };
      let cgBib = 0;
      let debate = 0;
      let cgOrg = 0;
      (evento.provas || []).forEach(function (p) {
        const bucket = tiebreakProvaBucket(p.titulo);
        if (!bucket) return;
        const pl = podium[p.id];
        const pts = gincanaPtsOnProva(pl, medalhas, gid);
        if (bucket === "cgBib") cgBib += pts;
        else if (bucket === "debate") debate += pts;
        else if (bucket === "cgOrg") cgOrg += pts;
      });
      out[gid] = {
        ou: m.ou || 0,
        pt: m.pt || 0,
        cgBiblia: cgBib,
        debate: debate,
        cgOrg: cgOrg,
      };
    });
    return out;
  }

  function computeTotals(evento, dados) {
    const pesos = (evento && evento.pesos) || {};
    const medalhas = (evento && evento.medalhas) || { ou: 0, pt: 0, br: 0 };
    const igrejas = Array.isArray(evento && evento.igrejas) ? evento.igrejas : [];
    const igrejaIds = new Set(
      igrejas.map(function (g) {
        return g && g.id;
      })
    );
    const participacao = (dados && dados.participacao) || {};
    const podium = (dados && dados.podium) || {};
    const gincTotals = gincanaPorIgreja(podium, medalhas, igrejaIds);
    const detalhes = [];
    for (let gi = 0; gi < igrejas.length; gi++) {
      const g = igrejas[gi];
      const gid = g.id;
      const row = participacao[gid] || {};
      const pPart = pontosParticipacao(row, pesos);
      const pPuni = pontosPunicoes(row, pesos);
      const pGinc = Number(gincTotals[gid] || 0);
      const pExt = pontuacaoExtra(row);
      const total = pPart + pPuni + pGinc + pExt;
      detalhes.push({
        igrejaId: gid,
        igreja: g.nome,
        participacao: pPart,
        punicoes: pPuni,
        gincana: pGinc,
        pontuacao_extra: pExt,
        total: total,
      });
    }
    const totais = detalhes.map(function (d) {
      return d.total;
    });
    const ranks = rankList(totais);
    const medalhasPorIgreja = contarMedalhasPorIgreja(podium, igrejaIds);
    const tiebreakByIgreja = buildTiebreakByIgreja(evento, dados);
    return {
      detalhes: detalhes,
      totais: totais,
      ranks: ranks,
      medalhasPorIgreja: medalhasPorIgreja,
      gincanaPorIgreja: gincTotals,
      tiebreakByIgreja: tiebreakByIgreja,
    };
  }

  function rankList(totais) {
    const n = totais.length;
    const out = new Array(n);
    for (let i = 0; i < n; i++) out[i] = 0;
    const order = [];
    for (let i = 0; i < n; i++) order.push(i);
    order.sort(function (a, b) {
      return totais[b] - totais[a];
    });
    for (let j = 0; j < n; j++) {
      const i = order[j];
      if (j === 0) out[i] = 1;
      else {
        const prev = order[j - 1];
        out[i] = totais[i] === totais[prev] ? out[prev] : j + 1;
      }
    }
    return out;
  }

  function contarMedalhasPorIgreja(podium, igrejaIds) {
    const out = {};
    igrejaIds.forEach(function (iid) {
      out[iid] = { ou: 0, pt: 0, br: 0 };
    });
    Object.values(podium).forEach(function (places) {
      if (!places || typeof places !== "object") return;
      ["ou", "pt", "br"].forEach(function (key) {
        const entry = places[key];
        if (!entry || typeof entry !== "object") return;
        const iid = entry.igrejaId;
        if (iid && out[iid] && out[iid][key] != null) out[iid][key] += 1;
      });
    });
    return out;
  }

  function classificacaoOrdenada(detalhes, ranks, tiebreakByIgreja) {
    const tb = tiebreakByIgreja || {};
    const emptyTb = { ou: 0, pt: 0, cgBiblia: 0, debate: 0, cgOrg: 0 };
    const indexed = [];
    for (let i = 0; i < detalhes.length; i++) indexed.push(i);
    indexed.sort(function (a, b) {
      const da = detalhes[a];
      const db = detalhes[b];
      let diff = db.total - da.total;
      if (diff !== 0) return diff;
      const ida = da.igrejaId;
      const idb = db.igrejaId;
      const xa = tb[ida] || emptyTb;
      const xb = tb[idb] || emptyTb;
      diff = xb.ou - xa.ou;
      if (diff !== 0) return diff;
      diff = xb.pt - xa.pt;
      if (diff !== 0) return diff;
      diff = xb.cgBiblia - xa.cgBiblia;
      if (diff !== 0) return diff;
      diff = xb.debate - xa.debate;
      if (diff !== 0) return diff;
      diff = xb.cgOrg - xa.cgOrg;
      if (diff !== 0) return diff;
      return String(da.igreja || "").localeCompare(String(db.igreja || ""), "pt", {
        sensitivity: "base",
        numeric: true,
      });
    });
    return indexed.map(function (i) {
      const row = Object.assign({}, detalhes[i]);
      row.posicao = ranks[i];
      return row;
    });
  }

  function validateEventoMinimal(evento) {
    const errs = [];
    if (!evento || typeof evento !== "object") return ["Evento deve ser um objeto."];
    ["meta", "pesos", "medalhas", "igrejas", "provas"].forEach(function (k) {
      if (evento[k] == null) errs.push("Falta chave: " + k);
    });
    if (!Array.isArray(evento.igrejas)) errs.push("igrejas deve ser uma lista.");
    if (!Array.isArray(evento.provas)) errs.push("provas deve ser uma lista.");
    if (errs.length) return errs;
    const ids = evento.igrejas.map(function (g) {
      return g && g.id;
    });
    if (new Set(ids).size !== ids.length) errs.push("IDs de igrejas duplicados");
    const pids = evento.provas.map(function (p) {
      return p && p.id;
    });
    if (new Set(pids).size !== pids.length) errs.push("IDs de provas duplicados");
    return errs;
  }

  function emptyDadosTemplate(igrejaIds, provaIds) {
    const part = {};
    igrejaIds.forEach(function (gid) {
      part[gid] = {
        inscricao: true,
        pontualidade: true,
        mr_total: 0,
        mr_camisa: 0,
        mr_biblia: 0,
        visitantes: 0,
        animacao: false,
        mau_comportamento: false,
        pontuacao_extra: 0,
      };
    });
    const pod = {};
    provaIds.forEach(function (pid) {
      pod[pid] = {
        ou: { igrejaId: null, competidor: "" },
        pt: { igrejaId: null, competidor: "" },
        br: { igrejaId: null, competidor: "" },
      };
    });
    return { participacao: part, podium: pod };
  }

  function avisosPodiumDuplicado(podium) {
    const avisos = [];
    Object.keys(podium).forEach(function (provaId) {
      const places = podium[provaId];
      if (!places || typeof places !== "object") return;
      const seen = new Set();
      ["ou", "pt", "br"].forEach(function (key) {
        const iid = places[key] && places[key].igrejaId;
        if (!iid) return;
        if (seen.has(iid))
          avisos.push("Prova «" + provaId + "»: igreja repetida em mais de um lugar.");
        seen.add(iid);
      });
    });
    return avisos;
  }

  global.ConclaveEngine = {
    computeTotals: computeTotals,
    rankList: rankList,
    classificacaoOrdenada: classificacaoOrdenada,
    validateEventoMinimal: validateEventoMinimal,
    emptyDadosTemplate: emptyDadosTemplate,
    avisosPodiumDuplicado: avisosPodiumDuplicado,
    contarMedalhasPorIgreja: contarMedalhasPorIgreja,
    // Helpers expostos para reuso por integrações externas, evitando
    // reimplementação divergente das regras de pontuação/punição.
    pontosParticipacao: pontosParticipacao,
    pontosPunicoes: pontosPunicoes,
    pontuacaoExtra: pontuacaoExtra,
    truthy: truthy,
    num: num,
  };
})(typeof window !== "undefined" ? window : globalThis);
